{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2e128c09-f3ff-410e-9699-df5abdf0e6ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# install.packages(\"dior.tar.gz\", repos = NULL, type = \"source\")\n",
    "\n",
    "# install.packages(\"IRkernel\")\n",
    "# IRkernel::installspec(user = T)\n",
    "\n",
    "\n",
    "\n",
    "# install.packages(\"dior-main.zip\", repos = NULL, type = \"source\")\n",
    "\n",
    "# > install.packages(\"dior.tar.gz\", repos = NULL, type = \"source\")\n",
    "# ERROR: dependencies 'Seurat', 'hdf5r', 'SingleCellExperiment', 'SummarizedExperiment', 'Hmisc' are not available for package 'dior'\n",
    "# * removing '/opt/conda/lib/R/library/dior'\n",
    "# Warning message:\n",
    "# In install.packages(\"d\n",
    "\n",
    "\n",
    "# BiocManager::install(\"Hmisc\")\n",
    "\n",
    "\n",
    "# #docker images里用这个\n",
    "# #devtools::install_github('IRkernel/IRkernel')\n",
    "# install.packages(\"IRkernel\")\n",
    "# IRkernel::installspec(user = T)\n",
    "\n",
    "# name = 'ir40', displayname = 'R 4.0', \n",
    "# ————————————————\n",
    "# 版权声明：本文为CSDN博主「闻人玲珑」的原创文章，遵循CC 4.0 BY-SA版权协议，转载请附上原文出处链接及本声明。\n",
    "# 原文链接：https://blog.csdn.net/weixin_41242128/article/details/113524475\n",
    "\n",
    "\n",
    "# conda install -c r r-base=4.1 r-essentials  r-devtools\n",
    "\n",
    "\n",
    "\n",
    "# #######################\n",
    "# > BiocManager::install(\"Hmisc\")\n",
    "# arning: unable to access index for repository https://bioconductor.org/packages/3.14/workflows/src/contrib:\n",
    "#   download from 'https://bioconductor.org/packages/3.14/workflows/src/contrib/PACKAGES' failed\n",
    "# Bioconductor version 3.14 (BiocManager 1.30.16), R 4.1.2 (2021-11-01)\n",
    "# Installing package(s) 'Hmisc'\n",
    "\n",
    "# also installing the dependency ‘htmlTable’\n",
    "\n",
    "# trying URL 'https://cloud.r-project.org/src/contrib/htmlTable_2.4.1.tar.gz'\n",
    "# Content type 'application/x-gzip' length 439117 bytes (428 KB)\n",
    "# ==================================================\n",
    "# downloaded 428 KB\n",
    "\n",
    "# trying URL 'https://cloud.r-project.org/src/contrib/Hmisc_4.7-1.tar.gz'\n",
    "# Content type 'application/x-gzip' length 792617 bytes (774 KB)\n",
    "# ==================================================\n",
    "# downloaded 774 KB\n",
    "#  hoeffd.o init.o jacklins.o largrec.o mChoice.o maxempr.o nstr.o ranksort.o rcorr.o string_box.o wclosest.o -lgfortran -lm -lgomp -lquadmath -lpthread -L/home/zhulab/miniconda3/envs/R4/lib/R/lib -lR\n",
    "# installing to /home/zhulab/miniconda3/envs/R4/lib/R/library/00LOCK-Hmisc/00new/Hmisc/libs\n",
    "# ** R\n",
    "# ** inst\n",
    "# ** byte-compile and prepare package for lazy loading\n",
    "# ** help\n",
    "# *** installing help indices\n",
    "# ** building package indices\n",
    "# ** testing if installed package can be loaded from temporary location\n",
    "# ** checking absolute paths in shared objects and dynamic libraries\n",
    "# ** testing if installed package can be loaded from final location\n",
    "# ** testing if installed package keeps a record of temporary installation path\n",
    "# * DONE (Hmisc)\n",
    "# >\n",
    "# >\n",
    "# > install.packages(\"dior.tar.gz\", repos = NULL, type = \"source\")\n",
    "# * installing *source* package ‘dior’ ...\n",
    "# ** using staged installation\n",
    "# ** R\n",
    "# ** inst\n",
    "# ** byte-compile and prepare package for lazy loading\n",
    "# Save workspace image? [y/n/c]: n\n",
    "# (R4) zhulab@DESKTOP-574GPKI:/mnt/e/yyp/codeproject/format$ cd ../\n",
    "# (R4) zhulab@DESKTOP-574GPKI:/mnt/e/yyp/codeproject$ jupyter notebook\n",
    "# data='YTAsce.all_qc.rds'\n",
    "# data=readRDS(data)\n",
    "# file='./YTAsce.all_qc_tmp.h5'\n",
    "# dior:::write_h5(data,file=file)\n",
    "\n",
    "\n",
    "\n",
    "# dior:::read.rds 可以用 假如报错 要安装 getopt 包 进行正确传参\n",
    "\n",
    "# #####一般情况下 安装 dior 和 diopy  就可以了，其他代码可以修改  增强对数据的理解，\n",
    "# 它这里中间的文件就是特定格式的tmp.h5 文件  可以转化为三种格式 ，见 scdior 的说明\n",
    "\n",
    "# seurat 的data 对应 scanpy   raw.x   ，scaledata 对应X ;反过来scanpy   raw.x  默认对应了  seurat 的   data(报错counts),所以要检查数据。\n",
    "\n",
    "\n",
    "# https://www.jianshu.com/p/9e0fb7c2c43f\n",
    "\n",
    "\n",
    "\n",
    "# def read_rds(file: Union[str, None] = None,\n",
    "#              object_type:str = 'seurat',\n",
    "#              assay_name: str = 'RNA'\n",
    "#             ) -> anndata.AnnData:  这个函数可以读rds 文件转 h5ad  也可以输入tmp.h5 文件文件转 h5ad\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f957870b-f4da-46c6-89cd-774d4c18fa19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seurat object H5 IO\n",
    "\n",
    "# library\n",
    "library(Seurat)\n",
    "library(hdf5r)\n",
    "library(Matrix)\n",
    "library(Hmisc)\n",
    "\n",
    "#' seurat insert the dimr information\n",
    "#'\n",
    "#' @param seurat The seruat object\n",
    "#' @param seurat_list The list will be inserted to the seurat\n",
    "#'\n",
    "seurat.insert_dimr_ <- function(seurat, seurat_list){\n",
    "  for(dr in names(seurat_list[['dimR']])){\n",
    "    seurat@reductions[[dr]] <- seurat_list[['dimR']][[dr]]\n",
    "    rownames(seurat@reductions[[dr]]@cell.embeddings) = rownames(seurat[[]])\n",
    "  }\n",
    "  return(seurat)\n",
    "}\n",
    "\n",
    "#' seurat insert the spatial information\n",
    "#'\n",
    "#' @param seurat The seruat object\n",
    "#' @param seurat_list The list will be inserted to the seurat\n",
    "#'\n",
    "seurat.insert_spatial_ <- function(seurat, seurat_list){\n",
    "  seurat@images <- seurat_list[['spatial']]\n",
    "  return(seurat)\n",
    "}\n",
    "\n",
    "#' seurat insert the graphs information\n",
    "#'\n",
    "#' @param seurat The seruat object\n",
    "#' @param seurat_list The list will be inserted to the seurat\n",
    "#'\n",
    "seurat.insert_graphs_ <- function(seurat, seurat_list){\n",
    "  for(g in names(seurat_list[['graphs']])){\n",
    "    rownames(seurat_list[['graphs']][[g]]) <- rownames(seurat[[]])\n",
    "    colnames(seurat_list[['graphs']][[g]]) <- rownames(seurat[[]])\n",
    "    seurat@graphs[[g]] <- seurat_list[['graphs']][[g]]\n",
    "  }\n",
    "  return(seurat)\n",
    "}\n",
    "\n",
    "#' seurat insert the uns information\n",
    "#'\n",
    "#' @param seurat The seruat object\n",
    "#' @param seurat_list The list will be inserted to the seurat\n",
    "#'\n",
    "seurat.insert_uns_ <- function(seurat, seurat_list){\n",
    "  seurat@misc <- seurat_list[['uns']]\n",
    "  return(seurat)\n",
    "}\n",
    "\n",
    "#' seurat insert the layers information\n",
    "#'\n",
    "#' @param seurat The seruat object\n",
    "#' @param seurat_list The list will be inserted to the seurat\n",
    "#'\n",
    "seurat.insert_layers_ <- function(seurat, seurat_list){\n",
    "  for(l in names(seurat_list[['layers']])){\n",
    "    rownames(seurat_list[['layers']][[l]]) <- rownames(seurat_list[['var']][['X']])\n",
    "    colnames(seurat_list[['layers']][[l]]) <- rownames(seurat[[]])\n",
    "    seurat@assays[[l]] <- Seurat::CreateAssayObject(counts = seurat_list[['layers']][[l]])\n",
    "  }\n",
    "  return(seurat)\n",
    "}\n",
    "\n",
    "#' H5 to Seuart object\n",
    "#'\n",
    "#' Read h5 and converted h5 to the seurat object\n",
    "#' @param file The h5 file\n",
    "#' @param assay.name 'assay.name' is used to flag the data type, and the default is \"RNA\", meaning this is scRNA-seq data.\n",
    "#' @importFrom hdf5r H5File h5attr\n",
    "#'\n",
    "seurat_read_h5 <- function(file=NULL, assay.name = NULL){\n",
    "  if(!file.exists(file)){\n",
    "    stop('No such file or directory')\n",
    "  }\n",
    "  h5 <- H5File$new(filename = file, mode = 'r')\n",
    "  tryCatch({\n",
    "    data <- h5_to_seurat(h5 = h5, assay.name = assay.name)\n",
    "  },\n",
    "  error = function(e) {\n",
    "    print(e)\n",
    "  },\n",
    "  finally = {\n",
    "    h5$close_all()\n",
    "  }\n",
    "  )\n",
    "  return(data)\n",
    "}\n",
    "\n",
    "#' H5 to the seurat\n",
    "#'\n",
    "#' H5 file is converted to the seurat obejct\n",
    "#' @param h5 The h5 file in R\n",
    "#' @param assay.name 'assay.name' is used to flag the data type, and the default is \"RNA\", meaning this is scRNA-seq data.\n",
    "#' @importFrom hdf5r H5File h5attr\n",
    "#'\n",
    "h5_to_seurat <- function(h5, assay.name){\n",
    "  options(warn = -1)\n",
    "  if(h5attr(h5, 'assay_name') == assay.name){\n",
    "    seurat_list <- list()\n",
    "    for(h in names(h5)){\n",
    "      seurat_list[[h]] <- switch(h, data = to_data_(h5),\n",
    "                                 obs = to_obs_(h5),\n",
    "                                 var = to_var_(h5),\n",
    "                                 dimR = seurat.to_dimr_(h5, assay.name=assay.name),\n",
    "                                 layers = to_layers_(h5),\n",
    "                                 graphs = to_graphs_(h5),\n",
    "                                 spatial = seurat.to_spatial_(h5),\n",
    "                                 uns = to_uns_(h5))\n",
    "    }\n",
    "  }\n",
    "  #--- create the seurat object: add the counts or norm data\n",
    "  if(assay.name == 'spatial'){\n",
    "    assay.name <- capitalize(assay.name)\n",
    "  }\n",
    "  rownames(seurat_list[['data']][['X']]) <- rownames(seurat_list[['var']][['X']])\n",
    "  colnames(seurat_list[['data']][['X']]) <- rownames(seurat_list[['obs']])\n",
    "  if(all(c('X', 'rawX') %in% names(seurat_list[['data']]))){\n",
    "    rownames(seurat_list[['data']][['rawX']]) <-  rownames(seurat_list[['var']][['rawX']])\n",
    "    colnames(seurat_list[['data']][['rawX']]) <- rownames(seurat_list[['obs']])\n",
    "    seurat <- Seurat::CreateSeuratObject(counts = seurat_list[['data']][['rawX']], assay = assay.name)\n",
    "    seurat@assays[[assay.name]]@scale.data <- seurat_list[['data']][['X']]\n",
    "    seurat@assays[[assay.name]]@meta.features <- seurat_list[['var']][['rawX']]\n",
    "  }else{\n",
    "    seurat <- Seurat::CreateSeuratObject(counts = seurat_list[['data']][['X']], assay = assay.name)\n",
    "    seurat@assays[[assay.name]]@meta.features <- seurat_list[['var']][['X']]\n",
    "  }\n",
    "  seurat@meta.data <- seurat_list[['obs']]\n",
    "  for(sl in names(seurat_list)){\n",
    "    seurat <- switch(sl,\n",
    "                     data = seurat,\n",
    "                     obs = seurat,\n",
    "                     var = seurat,\n",
    "                     dimR = seurat.insert_dimr_(seurat, seurat_list),\n",
    "                     layers = seurat.insert_layers_(seurat, seurat_list),\n",
    "                     graphs = seurat.insert_graphs_(seurat, seurat_list),\n",
    "                     spatial = seurat.insert_spatial_(seurat, seurat_list),\n",
    "                     uns = seurat.insert_uns_(seurat, seurat_list))\n",
    "  }\n",
    "  return(seurat)\n",
    "}\n",
    "\n",
    "# --- seurat write the h5 file\n",
    "\n",
    "#' The seurat object to h5\n",
    "#'\n",
    "#' The Seurat object is converted to the h5 file.\n",
    "#' @param seurat The seurat object.\n",
    "#' @param file The h5 file.\n",
    "#' @param assay.name 'assay.name' is used to flag the data type. Defualt is \"RNA\". Available options are:\n",
    "#' \\itemize{\n",
    "#'   \\item \"RNA\": this is the scRNA-seq data.\n",
    "#'   \\item \"spatial\": this is the spatial data.}\n",
    "#' @param save.graphs Default is TRUE , determing whether to save the graph(cell-cell similarity network).\n",
    "#'   seurat graph is different from scanpy graph. Their relationship are set {\"distances\": \"knn\", \"connectivities\": \"snn\"} roughly.\n",
    "#' @param save.scale Default is FALSE, determint whether to save the scale.data(dense matrix)\n",
    "#' @importFrom hdf5r H5File h5attr\n",
    "#' @export\n",
    "#'\n",
    "seurat_write_h5 <- function(seurat = NULL, file = NULL, assay.name = NULL, save.graphs = TRUE, save.scale = FALSE){\n",
    "  if(is.null(file)){\n",
    "    stop('No such file or directory')\n",
    "  }\n",
    "  if(class(seurat) != 'Seurat'){\n",
    "    stop('object ', substitute(seurat), ' class is not Seurat object')\n",
    "  }\n",
    "  h5 <- H5File$new(filename = file, mode = 'w')\n",
    "  tryCatch({\n",
    "    seurat_to_h5(seurat = seurat, h5 = h5, assay.name = assay.name, save.graphs = save.graphs, save.scale = save.scale)\n",
    "    h5attr(h5, 'assay_name') <- assay.name\n",
    "  },\n",
    "  error = function(e) {\n",
    "    print(e)\n",
    "  },\n",
    "  finally = {\n",
    "    h5$close_all()\n",
    "  }\n",
    "  )\n",
    "}\n",
    "\n",
    "\n",
    "#' Seurat is converted to h5 file\n",
    "#'\n",
    "#' Seurat object is converted to h5 file.\n",
    "#' @param seurat The seurat object.\n",
    "#' @param h5 The h5 file in R.\n",
    "#' @param assay.name 'assay.name' is used to flag the data type. Defualt is \"RNA\". Available options are:\n",
    "#' \\itemize{\n",
    "#'   \\item \"RNA\": this is the scRNA-seq data.\n",
    "#'   \\item \"spatial\": this is the spatial data.}\n",
    "#' @param save.graphs Default is TRUE , determing whether to save the graph(cell-cell similarity network).\n",
    "#'   seurat graph is different from scanpy graph. Their relationship are set {\"distances\": \"knn\", \"connectivities\": \"snn\"} roughly.\n",
    "#' @param save.scale Default is FALSE, determint whether to save the scale.data(dense matrix)\n",
    "#' @importFrom hdf5r H5File h5attr\n",
    "#' @importFrom Hmisc capitalize\n",
    "#'\n",
    "seurat_to_h5 <- function(seurat=NULL, h5=NULL, assay.name = NULL, save.graphs = TRUE, save.scale = FALSE){\n",
    "  data = h5$create_group('data')\n",
    "  var = h5$create_group('var')\n",
    "  if(assay.name == 'spatial'){\n",
    "    seurat_spatial_to_h5(data=seurat, h5 = h5,gr_name = assay.name)\n",
    "    assay.name <- capitalize(assay.name)\n",
    "  }\n",
    "  if(assay.name %in% names(slot(object = seurat, name = 'assays'))){\n",
    "    #--- data\n",
    "    slot_assay <- slot(object = seurat, name = 'assays')[[assay.name]]\n",
    "    if(length(slot_assay@scale.data)>0){\n",
    "      if(save.scale){\n",
    "        n1 = 'rawX'\n",
    "        sdata = 'X'\n",
    "      }else{\n",
    "        n1 = 'X'\n",
    "        sdata = NULL\n",
    "      }\n",
    "    }else{\n",
    "      n1 = 'X'\n",
    "      sdata = NULL\n",
    "    }\n",
    "    df_to_h5(df = slot(object = slot_assay, name = 'meta.features'), h5 = var, gr_name = n1)\n",
    "    if(all(slot_assay@data[,1] == slot_assay@counts[,1])){\n",
    "      matrix_to_h5(mat = slot(object = slot_assay, name = 'counts'), h5 = data, gr_name = n1)\n",
    "    }else{\n",
    "      layers = h5$create_group('layers')\n",
    "      matrix_to_h5(mat = slot(object = slot_assay, name = 'counts'), h5 = layers, gr_name= 'counts')\n",
    "      matrix_to_h5(mat = slot(object = slot_assay, name = 'data'), h5 = data, gr_name = n1)\n",
    "    }\n",
    "    if(!is.null(sdata)){\n",
    "      matrix_to_h5(mat = slot(object = slot_assay, name = 'scale.data'), h5 = data, gr_name = sdata)\n",
    "      var1 = slot(object = slot_assay, name = 'meta.features')\n",
    "      df_to_h5(df = var1[rownames(slot(object = slot_assay, name = 'scale.data')), ], h5 = var, gr_name = sdata)\n",
    "    }\n",
    "    #--- save the cell annotation\n",
    "    df_to_h5(df = slot(object = seurat, name = 'meta.data'), h5 = h5, gr_name = 'obs')\n",
    "    #--- save the dimension reduction\n",
    "    if(length(seurat@reductions)>0){\n",
    "      dimR <- h5$create_group('dimR')\n",
    "      dim.r <- slot(object = seurat, name = 'reductions')\n",
    "      for(d in names(dim.r)){\n",
    "        D = toupper(d)\n",
    "        dimR[[D]] <- t(slot(dim.r[[d]],'cell.embeddings'))\n",
    "      }\n",
    "    }\n",
    "    #--- save the graphs\n",
    "    if(save.graphs){\n",
    "      if(length(seurat@graphs)>0){\n",
    "        graph_df <- slot(object = seurat, 'graphs')\n",
    "        if(length(grep(assay.name, names(graph_df))) == 2){\n",
    "          graphs <- h5$create_group('graphs')\n",
    "          gra_list <- list()\n",
    "          gra_list[[paste0(assay.name, '_nn')]] <- 'knn'\n",
    "          gra_list[[paste0(assay.name, '_snn')]] <- 'snn'\n",
    "          for(g in names(graph_df)){\n",
    "            matrix_to_h5(mat=graph_df[[g]], h5 = graphs, gr_name = gra_list[[g]])\n",
    "          }\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    # save the layer\n",
    "    if(length(setdiff(names(seurat@assays), assay.name)) > 0){\n",
    "      if(! 'layers' %in% names(h5)){\n",
    "        layers = h5$create_group('layers')\n",
    "      }\n",
    "      for(s in setdiff(names(seurat@assays), assay.name)){\n",
    "        slot_layer <- slot(object = seurat, name = 'assays')[[s]]\n",
    "        matrix_to_h5(mat = slot_layer@data, h5 = layers, gr_name = s)\n",
    "\n",
    "      }\n",
    "    }\n",
    "    #--- save the metadata colors\n",
    "    if('uns' %in% names(seurat@misc)){\n",
    "      uns <- h5$create_group('uns')\n",
    "      for(colr in grep('colors', seurat@misc[['uns']], value = TRUE)){\n",
    "        uns[[colr]] <- seurat@misc[['uns']][[colr]]\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  else{\n",
    "    stop('Please enter the correct assay name')\n",
    "  }\n",
    "  return('Done!')\n",
    "}\n",
    "\n",
    "\n",
    "#' create the VisiumV1 class inherting the package Seurat 'SpatialImage' class\n",
    "#' @import Seurat\n",
    "#' @importFrom methods new\n",
    "VisiumV1 <- setClass(\n",
    "  Class = 'VisiumV1',\n",
    "  contains = 'SpatialImage',\n",
    "  slots = list(\n",
    "    'image' = 'array',\n",
    "    'scale.factors' = 'scalefactors',\n",
    "    'coordinates' = 'data.frame',\n",
    "    'spot.radius' = 'numeric'\n",
    "  )\n",
    ")\n",
    "\n",
    "\n",
    "#' The spatial meassage to the h5 file(Seurat)\n",
    "#'\n",
    "#' @param data The seruat object\n",
    "#' @param h5 The h5 file\n",
    "#' @param gr_name 'gr.name' must be the 'spatial' for saving the spatial message.\n",
    "#' @importFrom hdf5r H5File h5attr\n",
    "#'\n",
    "seurat_spatial_to_h5 <- function(data, h5, gr_name = 'spatial'){\n",
    "  h5spa <- h5$create_group(gr_name)\n",
    "  for(sample_id in names(data@images)){\n",
    "    sid_h5 <-  h5spa$create_group(sample_id)\n",
    "    #--- save the image for the lowres\n",
    "    sid_image_h5 <- sid_h5$create_group('image')\n",
    "    sid_image_h5[['lowres']] <-  aperm(slot(data@images[[sample_id]], 'image'))\n",
    "    #--- save the scale factor\n",
    "    sid_scalefactors_h5 <- sid_h5$create_group('scalefactors')\n",
    "    sf <- slot(data@images[[sample_id]], 'scale.factors')\n",
    "    v1 <- c('spot_diameter_fullres','fiducial_diameter_fullres','tissue_hires_scalef', 'tissue_lowres_scalef')\n",
    "    for(k in names(sf)){\n",
    "      sid_scalefactors_h5[[grep(k,v1, value = TRUE)]] <- sf[[k]]\n",
    "    }\n",
    "    #--- save the coor\n",
    "    coor_df = slot(data@images[[sample_id]], 'coordinates')\n",
    "    colnames(coor_df) <- c('in_tissue', 'array_row','array_col', 'image_2','image_1')\n",
    "    df_to_h5(df = coor_df[,c('in_tissue', 'array_row','array_col','image_1','image_2')], h5 = sid_h5, gr_name = 'coor')\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "#' The h5 group spatial to the spatial message\n",
    "#'\n",
    "#' @param h5spa The h5 group for the spatial\n",
    "#' @importFrom methods new slot\n",
    "#' @return The spatial message\n",
    "#'\n",
    "h5_to_spatial <- function(h5spa){\n",
    "  spatial_list<- list()\n",
    "  for(sid in names(h5spa)){\n",
    "    spatial_sid_list <- list()\n",
    "    sid_h5 <- h5spa[[sid]]\n",
    "    for(me in names(sid_h5)){\n",
    "      if('image' == me){\n",
    "        spatial_sid_list[[me]] <- aperm(sid_h5[[me]][['lowres']][,,])\n",
    "      }\n",
    "      if('scalefactors' == me){\n",
    "        sf_list <- list()\n",
    "        v1 = c(\"spot\",\"fiducial\",\"hires\",\"lowres\")\n",
    "        for(sf in v1){\n",
    "          sf_list[[sf]] <- sid_h5[[me]][[grep(sf,names(sid_h5[[me]]), value = TRUE)]][]\n",
    "        }\n",
    "        sf_o <- Seurat::scalefactors(spot = sf_list$spot, fiducial = sf_list$fiducial, hires = sf_list$hires, lowres = sf_list$lowres)\n",
    "        spatial_sid_list[[me]] <- sf_o\n",
    "      }\n",
    "      if('coor' == me){\n",
    "        coor_df <- h5_to_df(sid_h5[['coor']])\n",
    "        coor_df <- coor_df[,c('in_tissue','array_row','array_col','image_2','image_1')]\n",
    "        colnames(coor_df)<- c('tissue','row','col','imagerow','imagecol')\n",
    "        spatial_sid_list[[me]] <- coor_df\n",
    "      }\n",
    "    }\n",
    "    unnormalized.radius <- spatial_sid_list$scalefactors$fiducial * spatial_sid_list$scalefactors$lowres\n",
    "    spot.radius <- unnormalized.radius/max(dim(x = spatial_sid_list$image))\n",
    "    spatial_list[[sid]] <- new(Class = \"VisiumV1\", image = spatial_sid_list$image, scale.factors = spatial_sid_list$scalefactors,\n",
    "                               coordinates = spatial_sid_list$coor,spot.radius = spot.radius)\n",
    "    spatial_list[[sid]]@key <- names(h5spa)\n",
    "    spatial_list[[sid]]@assay <- 'Spatial'\n",
    "  }\n",
    "  return(spatial_list)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "98b19350-11d8-402d-9620-2b37c714f5fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- SingleCellExperiment(sce) H5 IO\n",
    "\n",
    "# library\n",
    "library(hdf5r)\n",
    "library(Matrix)\n",
    "library(SingleCellExperiment)\n",
    "\n",
    "# --- SingleCellExperiment write h5 file\n",
    "\n",
    "#' The singlecellexperiment is converted to h5 file\n",
    "#'\n",
    "#' Singlecellexperiment object is converted to h5 file.\n",
    "#' @param sce The singlecellexperiment object.\n",
    "#' @param file The h5 file.\n",
    "#' @param assay.name The 'assay.name' is used to flag the data type, and the default is \"RNA\", meaning this is scRNA-seq data.\n",
    "#' @importFrom hdf5r H5File h5attr\n",
    "#'\n",
    "sce_write_h5 <- function(sce = NULL, file = NULL, assay.name = NULL){\n",
    "  if(is.null(file)){\n",
    "    stop('No such file or directory')\n",
    "  }\n",
    "  if(class(sce) != 'SingleCellExperiment'){\n",
    "    stop('object ', substitute(sce), ' class is not SingleCellExperiment object')\n",
    "  }\n",
    "  h5 <- H5File$new(filename = file, mode = 'w')\n",
    "  tryCatch({\n",
    "    sce_to_h5(sce = sce, h5 = h5, assay.name = assay.name)\n",
    "    h5attr(h5, 'assay_name') <- assay.name\n",
    "  },\n",
    "  error = function(e){\n",
    "    print(e)\n",
    "  },\n",
    "  finally = {\n",
    "    h5$close_all()\n",
    "  })\n",
    "}\n",
    "\n",
    "#' The singlecellexperiment is converted to h5 file\n",
    "#'\n",
    "#' Singlecellexperiment object is converted to h5 file.\n",
    "#' @param sce The singlecellexperiment object.\n",
    "#' @param h5 The h5 file in R.\n",
    "#' @param assay.name The 'assay.name' is used to flag the data type, and the default is \"RNA\", meaning this is scRNA-seq data.\n",
    "#' @importFrom hdf5r H5File h5attr\n",
    "#' @importFrom SummarizedExperiment assayNames assay colData rowData \"assayNames<-\"\n",
    "#' @importFrom SingleCellExperiment reducedDimNames reducedDim\n",
    "sce_to_h5 <- function(sce, h5, assay.name){\n",
    "  h5attr(h5, 'assay_name') <- assay.name\n",
    "  data <- h5$create_group('data')\n",
    "  var <- h5$create_group('var')\n",
    "  #--- save the matrix\n",
    "  if(length(assayNames(sce))> 1){\n",
    "    layers <- h5$create_group('layers')\n",
    "  }\n",
    "  if(!'X' %in% assayNames(sce)){\n",
    "    assayNames(sce)[1] <- 'X'\n",
    "    print(\"The first 'assayNames' defaults to 'X'\")\n",
    "  }\n",
    "  for(as in assayNames(sce)){\n",
    "    if('X' == as){\n",
    "      matrix_to_h5(mat = assay(sce, as), h5 = data, gr_name = as)\n",
    "    }else{\n",
    "      matrix_to_h5(mat = assay(sce, as), h5 = layers, gr_name = as)\n",
    "    }\n",
    "  }\n",
    "  #--- save cell annotation\n",
    "  df_to_h5(df = colData(sce), h5 = h5, gr_name = 'obs')\n",
    "  #--- save var\n",
    "  df_to_h5(df = rowData(sce), h5 = var, gr_name = 'X')\n",
    "  #--- save reduction dimension\n",
    "  if(length(reducedDimNames(sce))>0){\n",
    "    dimReduction <- h5$create_group('dimR')\n",
    "    for(d in reducedDimNames(sce)){\n",
    "      D = toupper(d)\n",
    "      dimReduction[[D]] <- t(reducedDim(sce, d))\n",
    "    }\n",
    "  }\n",
    "\n",
    "}\n",
    "\n",
    "# --- singlecellexperiment read the h5 file\n",
    "\n",
    "#' H5 to singlecellexperiment object\n",
    "#'\n",
    "#' Read h5 and converted h5 to the singlecellexperiment object\n",
    "#' @param file The h5 file\n",
    "#' @param assay.name 'assay.name' is used to flag the data type, and the default is \"RNA\", meaning this is scRNA-seq data.\n",
    "#' @importFrom hdf5r H5File h5attr\n",
    "#'\n",
    "sce_read_h5 <- function(file=NULL, assay.name = NULL){\n",
    "  if(!file.exists(file)){\n",
    "    stop('No such file or directory')\n",
    "  }\n",
    "  h5 <- H5File$new(filename = file, mode = 'r')\n",
    "  tryCatch({\n",
    "    data <- h5_to_sce(h5 = h5, assay.name = assay.name)\n",
    "  },\n",
    "  error = function(e) {\n",
    "    print(e)\n",
    "  },\n",
    "  finally = {\n",
    "    h5$close_all()\n",
    "  }\n",
    "  )\n",
    "  return(data)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "#' H5 to the singlecellexperiment\n",
    "#'\n",
    "#' H5 file is converted to the singlecellexperiment obejct\n",
    "#' @param h5 The h5 file in R\n",
    "#' @param assay.name 'assay.name' is used to flag the data type, and the default is \"RNA\", meaning this is scRNA-seq data.\n",
    "#' @param use.data using the scanpy the data.X slot\n",
    "#' @importFrom hdf5r H5File h5attr\n",
    "#'\n",
    "h5_to_sce <- function(h5, assay.name = 'RNA',use.data ='X'){\n",
    "  options(warn = -1)\n",
    "  if(h5attr(h5, 'assay_name') == assay.name){\n",
    "    sce_list <- list()\n",
    "    for(h in names(h5)){\n",
    "      # tran to the list\n",
    "      sce_list[[h]] <- switch(h, data = to_data_(h5),\n",
    "                              obs = to_obs_(h5),\n",
    "                              var = to_var_(h5),\n",
    "                              dimR = to_dimr_(h5),\n",
    "                              layers = to_layers_(h5),\n",
    "                              graphs = to_graphs_(h5),\n",
    "                              uns = to_uns_(h5))\n",
    "    }\n",
    "    if('dimR' %in% names(sce_list)){\n",
    "      dimR_list <- sce_list[['dimR']]\n",
    "      for(dr in names(dimR_list)){\n",
    "        rownames(sce_list[['dimR']][[dr]]) <- rownames(sce_list[['obs']])\n",
    "      }\n",
    "    }\n",
    "    if(use.data == 'rawX'){\n",
    "      ass <- sce_list[['data']][use.data]\n",
    "      rownames(ass[[use.data]]) <-  rownames(sce_list[['var']][[use.data]])\n",
    "      colnames(ass[[use.data]]) <- rownames(sce_list[['obs']])\n",
    "    }\n",
    "    if(use.data == 'X'){\n",
    "      ass <- list(X = sce_list[['data']][[use.data]])\n",
    "      rownames(ass[[use.data]]) <-  rownames(sce_list[['var']][[use.data]])\n",
    "      colnames(ass[[use.data]]) <- rownames(sce_list[['obs']])\n",
    "      if('layers' %in% names(sce_list)){\n",
    "        for(l in names(sce_list[['layers']])){\n",
    "          ass[[l]] <- sce_list[['layers']][[l]]\n",
    "        }\n",
    "        for(d in names(ass)){\n",
    "          rownames(ass[[d]]) <-  rownames(sce_list[['var']][[use.data]])\n",
    "          colnames(ass[[d]]) <- rownames(sce_list[['obs']])\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    # save the all dataset and doing\n",
    "    sce <- SingleCellExperiment::SingleCellExperiment(assays =ass,\n",
    "                                                      colData = sce_list[['obs']],\n",
    "                                                      rowData = sce_list[['var']][[use.data]],\n",
    "                                                      reducedDims = sce_list[['dimR']])\n",
    "\n",
    "  }\n",
    "  return(sce)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "09d24e42-117d-4b54-94a0-b187bbb56834",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---  base data IO\n",
    "\n",
    "# packages\n",
    "library(hdf5r)\n",
    "library(Matrix)\n",
    "library(Hmisc)\n",
    "\n",
    "#' Data frame to h5\n",
    "#'\n",
    "#' Data frame is converted to the h5 file\n",
    "#' @param df Data frame of cell annotation or gene annotation\n",
    "#' @param h5 The h5 file\n",
    "#' @param gr_name The group name represents the property of the data frame.\n",
    "#' @importFrom hdf5r H5File h5attr \"h5attr<-\" h5attr_names\n",
    "#' @export\n",
    "#'\n",
    "# df_to_h5 <- function(df, h5, gr_name=NULL){\n",
    "#   h5df <- h5$create_group(gr_name)\n",
    "#   h5df[['index']] = rownames(df)\n",
    "#   if(ncol(df)>0){\n",
    "#     h5df[['colnames']] = colnames(df)\n",
    "#   }\n",
    "#   # factor to levels,charactor to levels,logical to levels\n",
    "#   for(k in names(df)){\n",
    "#     if(is.factor(df[[k]])){\n",
    "#       e0 <- as.integer(df[[k]]) - 1L\n",
    "#       e0[is.na(e0)] <- -1 # -1 is na\n",
    "#       h5df[[k]]<- e0 # for 0 begin\n",
    "#\n",
    "#       h5df[[paste0(k,'_levels')]]<- levels(df[[k]])\n",
    "#       h5attr(h5df[[k]], 'origin_dtype') = 'category'\n",
    "#     }\n",
    "#     if(is.character(df[[k]])){\n",
    "#       str_to_lvl <- factor(df[[k]])\n",
    "#       e0 <- as.integer(str_to_lvl) - 1L\n",
    "#       e0[is.na(e0)] <- -1 # -1 is na\n",
    "#       h5df[[k]]<- e0\n",
    "#       h5df[[paste0(k,'_levels')]]<- levels(str_to_lvl)\n",
    "#       h5attr(h5df[[k]], 'origin_dtype') = 'string'\n",
    "#     }\n",
    "#     if(is.logical(df[[k]])){\n",
    "#       h5df[[k]] <- as.integer(df[[k]])\n",
    "#       h5attr(h5df[[k]], 'origin_dtype') = 'bool'\n",
    "#     }\n",
    "#     if(any(is.numeric(df[[k]]),is.integer(df[[k]]))){\n",
    "#       h5df[[k]] <- df[[k]]\n",
    "#       h5attr(h5df[[k]], 'origin_dtype') = 'number'\n",
    "#     }\n",
    "#   }\n",
    "# }\n",
    "\n",
    "df_to_h5 <- function(df, h5, gr_name=NULL){\n",
    "  h5df <- h5$create_group(gr_name)\n",
    "  cate_list <- list()\n",
    "  h5df[['index']] = rownames(df)\n",
    "  if(ncol(df)>0){\n",
    "    h5df[['colnames']] = colnames(df)\n",
    "  }\n",
    "  # factor to levels,charactor to levels,logical to levels\n",
    "  for(k in names(df)){\n",
    "    if(is.factor(df[[k]])){\n",
    "      e0 <- as.integer(df[[k]]) - 1L\n",
    "      e0[is.na(e0)] <- -1 # -1 is na\n",
    "      h5df[[k]]<- e0 # for 0 begin\n",
    "      cate_list[[k]] <- levels(df[[k]])\n",
    "      h5attr(h5df[[k]], 'origin_dtype') = 'category'\n",
    "    }\n",
    "    if(is.character(df[[k]])){\n",
    "      str_to_lvl <- factor(df[[k]])\n",
    "      e0 <- as.integer(str_to_lvl) - 1L\n",
    "      e0[is.na(e0)] <- -1 # -1 is na\n",
    "      h5df[[k]]<- e0\n",
    "      cate_list[[k]] <- levels(str_to_lvl)\n",
    "      h5attr(h5df[[k]], 'origin_dtype') = 'string'\n",
    "    }\n",
    "    if(is.logical(df[[k]])){\n",
    "      h5df[[k]] <- as.integer(df[[k]])\n",
    "      h5attr(h5df[[k]], 'origin_dtype') = 'bool'\n",
    "    }\n",
    "    if(any(is.numeric(df[[k]]),is.integer(df[[k]]))){\n",
    "      h5df[[k]] <- df[[k]]\n",
    "      h5attr(h5df[[k]], 'origin_dtype') = 'number'\n",
    "    }\n",
    "  }\n",
    "  if(length(names(cate_list))>0){\n",
    "    h5df_cate <- h5df$create_group('category')\n",
    "    for(ca in names(cate_list)){\n",
    "      h5df_cate[[ca]] <- cate_list[[ca]]\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "#' H5 to dataframe\n",
    "#'\n",
    "#' H5 gruop is converted to usable data frame including the observes annotation (cell annotation) and variables annotation (gene annotation)\n",
    "#' @param h5df The hdf5 group saved the data frame using list mode.\n",
    "#' @importFrom hdf5r H5File h5attr \"h5attr<-\" h5attr_names\n",
    "#' @return A data frame(annotation)\n",
    "#' @export\n",
    "#'\n",
    "h5_to_df <- function(h5df){\n",
    "  df_list <- list()\n",
    "  df_list[['index']] <- h5df[['index']][]\n",
    "  for(k in names(h5df)){\n",
    "    if(length(h5attr_names(h5df[[k]]))>0){\n",
    "      df_dtype <- h5attr(h5df[[k]], 'origin_dtype')\n",
    "      if(df_dtype == 'category'){\n",
    "        e0 <- h5df[[k]][]\n",
    "        e0[e0==-1] <- NA # -1 is na\n",
    "        e0 <- e0 + 1L\n",
    "        lvl <- h5df[['category']][[k]][]\n",
    "        df_list[[k]] <- structure(.Data = e0, .Label = lvl, class = 'factor')\n",
    "      }\n",
    "      if(df_dtype == 'string'){\n",
    "        e0 <- h5df[[k]][]\n",
    "        e0[e0==-1] <- NA # -1 is na\n",
    "        e0 <- e0 + 1L\n",
    "        lvl <- h5df[['category']][[k]][]\n",
    "        df_list[[k]] <- as.character(structure(.Data = e0, .Label = lvl, class = 'factor'))\n",
    "      }\n",
    "      if(df_dtype == 'bool'){\n",
    "        df_list[[k]] <- as.logical(h5df[[k]][])\n",
    "      }\n",
    "      if(df_dtype == 'number'){\n",
    "        df_list[[k]] <- h5df[[k]][]\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  df = as.data.frame(df_list, row.names = 'index', optional = TRUE)\n",
    "  if('colnames' %in% names(h5df)){\n",
    "    cnames <- h5df[['colnames']][]\n",
    "    if(length(cnames) >1){\n",
    "      df = df[,cnames]\n",
    "    }else{\n",
    "      df = df\n",
    "    }\n",
    "  }else{\n",
    "    df=df\n",
    "  }\n",
    "  return(df)\n",
    "}\n",
    "\n",
    "# h5_to_df <- function(h5df){\n",
    "#   df_list <- list()\n",
    "#   df_list[['index']] <- h5df[['index']][]\n",
    "#   for(k in names(h5df)){\n",
    "#     if(length(h5attr_names(h5df[[k]]))>0){\n",
    "#       df_dtype <- h5attr(h5df[[k]], 'origin_dtype')\n",
    "#       if(df_dtype == 'category'){\n",
    "#         e0 <- h5df[[k]][]\n",
    "#         e0[e0==-1] <- NA # -1 is na\n",
    "#         e0 <- e0 + 1L\n",
    "#         lvl <- h5df[[paste0(k,'_levels')]][]\n",
    "#         df_list[[k]] <- structure(.Data = e0, .Label = lvl, class = 'factor')\n",
    "#       }\n",
    "#       if(df_dtype == 'string'){\n",
    "#         e0 <- h5df[[k]][]\n",
    "#         e0[e0==-1] <- NA # -1 is na\n",
    "#         e0 <- e0 + 1L\n",
    "#         lvl <- h5df[[paste0(k,'_levels')]][]\n",
    "#         df_list[[k]] <- as.character(structure(.Data = e0, .Label = lvl, class = 'factor'))\n",
    "#       }\n",
    "#       if(df_dtype == 'bool'){\n",
    "#         df_list[[k]] <- as.logical(h5df[[k]][])\n",
    "#       }\n",
    "#       if(df_dtype == 'number'){\n",
    "#         df_list[[k]] <- h5df[[k]][]\n",
    "#       }\n",
    "#     }\n",
    "#   }\n",
    "#   df = as.data.frame(df_list, row.names = 'index', optional = TRUE)\n",
    "#   if('colnames' %in% names(h5df)){\n",
    "#     cnames <- h5df[['colnames']][]\n",
    "#     if(length(cnames) >1){\n",
    "#       df = df[,cnames]\n",
    "#     }else{\n",
    "#       df = df\n",
    "#     }\n",
    "#   }else{\n",
    "#     df=df\n",
    "#   }\n",
    "#   return(df)\n",
    "# }\n",
    "\n",
    "\n",
    "#' Matrix to H5 format\n",
    "#'\n",
    "#' The matrix including the dense matrix and sparse matrix is converted to the matrix in h5 format or is stored into the h5 file.\n",
    "#' @param mat The matrix object including matrix(R) and sparse matrix(Matrix package)\n",
    "#' @param h5 The H5 file name that we write in\n",
    "#' @param gr_name The h5 gorup store the matrix (dense matrix or sparse matrix)\n",
    "#' @param save.obs.name The rownames isn't be saved(FALSE by defualt)\n",
    "#' @param save.var.name The colnames isn't be saved(FALSE by defualt)\n",
    "#' @importFrom hdf5r H5File h5attr\n",
    "#' @importFrom methods slot\n",
    "#' @export\n",
    "#'\n",
    "matrix_to_h5 <- function(mat, h5, gr_name = NULL, save.obs.name = FALSE, save.var.name = FALSE){\n",
    "  if(!gr_name %in% names(h5)){\n",
    "    h5mat = h5$create_group(gr_name)\n",
    "  }\n",
    "  else{\n",
    "    h5mat = h5[[gr_name]]\n",
    "  }\n",
    "  if('dgCMatrix' %in% class(mat)){\n",
    "    h5mat[['values']] <- slot(object = mat, name = 'x')\n",
    "    h5mat[['indices']] <- slot(object = mat, name = 'i')\n",
    "    h5mat[['indptr']] <- slot(object = mat, name = 'p')\n",
    "    h5mat[['dims']] <- rev(slot(object = mat, name = 'Dim'))\n",
    "    if(save.obs.name & save.var.name){\n",
    "      h5mat[['var_names']] <- slot(object = mat, name = 'Dimnames')[[1]]\n",
    "      h5mat[['obs_names']] <- slot(object = mat, name = 'Dimnames')[[2]]\n",
    "    }\n",
    "    h5attr(h5mat, 'datatype') <- 'SparseMatrix'\n",
    "  }\n",
    "  else if('matrix' %in% class(mat)){\n",
    "    h5mat[['matrix']] <- mat\n",
    "    h5mat[['dims']] <- rev(dim(mat))\n",
    "    if(save.obs.name & save.var.name){\n",
    "      h5mat[['var_names']] <- slot(object = mat, name = 'Dimnames')[[1]]\n",
    "      h5mat[['obs_names']] <- slot(object = mat, name = 'Dimnames')[[2]]\n",
    "    }\n",
    "    h5attr(h5mat, 'datatype') <- 'Array'\n",
    "    warning(paste0(substitute(gr_name), ' is dense matrix'))#2\n",
    "  }\n",
    "  else if('Graph' %in% class(mat)){\n",
    "    h5mat[['values']] <- slot(object = mat, name = 'x')\n",
    "    h5mat[['indices']] <- slot(object = mat, name = 'i')\n",
    "    h5mat[['indptr']] <- slot(object = mat, name = 'p')\n",
    "    h5mat[['dims']] <- rev(slot(object = mat, name = 'Dim'))\n",
    "    h5attr(h5mat, 'datatype') <- 'SparseMatrix'\n",
    "  }else{\n",
    "    stop('The matrix type is wrong')\n",
    "  }\n",
    "}\n",
    "\n",
    "#' H5 to Matrix format\n",
    "#'\n",
    "#' H5 group is converted to usable matrice including dense matrices and sparse matrices.\n",
    "#' Dense matrices(by R) and sparse matrices(constructed by Matrix package)\n",
    "#' @param h5mat The name of the group the stores the matrix data in h5 file\n",
    "#' @param obs.name The observes names, such as cell names\n",
    "#' @param var.name The variables names, such as gene names\n",
    "#' @return dense matrix or sparse matrix\n",
    "#' @importFrom hdf5r H5File h5attr\n",
    "#' @export\n",
    "#'\n",
    "h5_to_matrix <-  function(h5mat, obs.name=NULL, var.name=NULL){\n",
    "  if(all(c('obs_names','var_names') %in% names(h5mat))){\n",
    "    obs.name = h5mat[['obs_names']][]\n",
    "    var.name = h5mat[['var_names']][]\n",
    "  }\n",
    "  if(h5attr(h5mat, 'datatype') == 'SparseMatrix'){\n",
    "    mat <- Matrix::sparseMatrix(i = h5mat[['indices']][],\n",
    "                                p = h5mat[['indptr']][],\n",
    "                                x = h5mat[['values']][],\n",
    "                                dims = rev(h5mat[['dims']][]),\n",
    "                                index1 = FALSE)\n",
    "    if(!is.null(obs.name) & !is.null(var.name)){\n",
    "      dimnames(mat) <- list(var.name, obs.name)\n",
    "    }\n",
    "    else{\n",
    "      warning('There are no dimnames in the sparse matrix')\n",
    "    }\n",
    "  }\n",
    "  else if(h5attr(h5mat, 'datatype') == 'Array'){\n",
    "    mat <- h5mat[['matrix']][,]\n",
    "    if(!is.null(obs.name) & !is.null(var.name)){\n",
    "      dimnames(mat) <- list(var.name, obs.name)\n",
    "    }\n",
    "    else{\n",
    "      warning('There are no dimnames in the matrix')\n",
    "    }\n",
    "  }\n",
    "  return(mat)\n",
    "}\n",
    "\n",
    "#' h5[['obs']] file transform obs information\n",
    "#'\n",
    "#' h5[['obs']] file transform obs information\n",
    "#' @param h5 The h5 file\n",
    "#'\n",
    "to_obs_ <- function(h5){\n",
    "  to_obs <- h5_to_df(h5df = h5[['obs']])\n",
    "  return(to_obs)\n",
    "}\n",
    "\n",
    "#' h5[['var]] file transform var information\n",
    "#'\n",
    "#' h5[['var]] file transform var information\n",
    "#' @param h5 The h5 file\n",
    "#'\n",
    "to_var_ <- function(h5){\n",
    "  to_var <- list()\n",
    "  var = h5[['var']]\n",
    "  for(v in names(var)){\n",
    "    to_var[[v]] <- h5_to_df(h5df = var[[v]])\n",
    "  }\n",
    "  return(to_var)\n",
    "}\n",
    "\n",
    "#' h5[['graphs']] file transform graphs information\n",
    "#'\n",
    "#' h5[['graphs']] file transform graphs information\n",
    "#' @param h5 The h5 file\n",
    "#'\n",
    "to_graphs_ <- function(h5){\n",
    "  to_graphs <- list()\n",
    "  graphs <- h5[['graphs']]\n",
    "  graphs_name <- list(knn = \"RNA_nn\", snn = \"RNA_snn\")\n",
    "  for(g in names(graphs)){\n",
    "    to_graphs[[graphs_name[[g]]]] = h5_to_matrix(h5mat = graphs[[g]])\n",
    "  }\n",
    "  return(to_graphs)\n",
    "}\n",
    "\n",
    "#' h5[['data']] file transform data information\n",
    "#'\n",
    "#' h5[['data']] file transform data information\n",
    "#' @param h5 The h5 file\n",
    "#'\n",
    "to_data_ <- function(h5){\n",
    "  to_data <- list()\n",
    "  data <- h5[['data']]\n",
    "  for(d in names(data)){\n",
    "    to_data[[d]] <- h5_to_matrix(h5mat = data[[d]])\n",
    "  }\n",
    "  return(to_data)\n",
    "}\n",
    "\n",
    "#' h5[['layers']] file transform the layer information\n",
    "#'\n",
    "#' h5[['layers']] file transform the layer information\n",
    "#' @param h5 The h5 file\n",
    "#'\n",
    "to_layers_ <- function(h5){\n",
    "  to_layers <- list()\n",
    "  layers <- h5[['layers']]\n",
    "  for(l in names(layers)){\n",
    "    to_layers[[l]] <- h5_to_matrix(h5mat = layers[[l]])\n",
    "  }\n",
    "  return(to_layers)\n",
    "}\n",
    "\n",
    "#' h5[['uns']] file transform the uns information\n",
    "#'\n",
    "#' h5[['uns']] file transform the uns information\n",
    "#' @param h5 The h5 file\n",
    "#'\n",
    "to_uns_ <- function(h5){\n",
    "  colors_list = list()\n",
    "  for(colr in grep('colors', names(h5[['uns']]), value = TRUE)){\n",
    "    colors_list[[colr]] <- h5[['uns']][[colr]][]\n",
    "  }\n",
    "  return(colors_list)\n",
    "}\n",
    "\n",
    "#' h5[['dimR']] file transform the dimr information\n",
    "#'\n",
    "#' h5[['dimR']] file transform the dimr information\n",
    "#' @param h5 The h5 file\n",
    "#'\n",
    "to_dimr_ <- function(h5){\n",
    "  to_dimr <- list()\n",
    "  dimr <- h5[['dimR']]\n",
    "  for(DR in names(dimr)){\n",
    "    to_dimr[[tolower(DR)]] <- t(dimr[[DR]][,])\n",
    "    colnames(to_dimr[[tolower(DR)]]) <-  paste0(DR, '_',1:ncol(to_dimr[[tolower(DR)]]))\n",
    "  }\n",
    "  return(to_dimr)\n",
    "}\n",
    "\n",
    "#' h5[['dimR']] file transform the dimr information of seurat object\n",
    "#'\n",
    "#' h5[['dimR']] file transform the dimr information of seurat object\n",
    "#' @param h5 The h5 file\n",
    "#' @param assay.name The assay.name\n",
    "#'\n",
    "seurat.to_dimr_ <- function(h5, assay.name){\n",
    "  if(assay.name == 'spatial'){\n",
    "    assay.name <- capitalize(assay.name)\n",
    "  }\n",
    "  to_dimr <- list()\n",
    "  dimr <- h5[['dimR']]\n",
    "  for(DR in names(dimr)){\n",
    "    to_dimr[[tolower(DR)]] <- Seurat::CreateDimReducObject(embeddings = t(dimr[[DR]][,]), key = paste0(DR, \"_\"), assay = assay.name)\n",
    "  }\n",
    "  return(to_dimr)\n",
    "}\n",
    "\n",
    "#' h5[['spatial']] file transforma the spatial information of seurat object\n",
    "#'\n",
    "#' h5[['spatial']] file transforma the spatial information of seurat object\n",
    "#' @param h5 The h5 file\n",
    "#'\n",
    "seurat.to_spatial_ <- function(h5){\n",
    "  to_spatial <- h5_to_spatial(h5[['spatial']])\n",
    "  return(to_spatial)\n",
    "}\n",
    "\n",
    "#--- read h5 file\n",
    "\n",
    "#' H5 to scRNAs-seq analysis object\n",
    "#'\n",
    "#' Read h5 and converted h5 to the scRNA-seq analysis object\n",
    "#' @param target.object Denotes which object to load. Defualt is \"seurat\". Available options are:\n",
    "#' \\itemize{\n",
    "#'   \\item \"seurat\": converted the h5 file to \"seurat object\".\n",
    "#'   \\item \"singlecellexperiment\": converted the h5 file to \"singlecellexperiment object\".\n",
    "#'   \\item \"monocle\": converted the h5 file to \"monocle3 object\".\n",
    "#' }\n",
    "#' @param file The h5 file\n",
    "#' @param assay.name 'assay.name' is used to flag the data type. Defualt is \"RNA\". Available options are:\n",
    "#' \\itemize{\n",
    "#'   \\item \"RNA\": this is the scRNA-seq data.\n",
    "#'   \\item \"spatial\": this is the spatial data.}\n",
    "#' @return The single cell analysis mainstream software\n",
    "#' @export\n",
    "#'\n",
    "read_h5 <- function(file, target.object = 'seurat', assay.name = 'RNA'){\n",
    "  if(target.object == 'seurat'){\n",
    "    data <- seurat_read_h5(file = file, assay.name = assay.name)\n",
    "  }\n",
    "  else if(target.object == 'singlecellexperiment'){\n",
    "    data <- sce_read_h5(file = file, assay.name = assay.name)\n",
    "  }\n",
    "  else{\n",
    "    stop('The output object must be specified')\n",
    "  }\n",
    "  return(data)\n",
    "}\n",
    "\n",
    "\n",
    "#--- write h5\n",
    "\n",
    "#' The scRNAs-seq analysis objec to H5\n",
    "#'\n",
    "#' Write h5 and  the scRNA-seq analysis object converted to h5\n",
    "#' @param data The scRNA-seq analysis object data.\n",
    "#' @param file The h5 file\n",
    "#' @param object.type Denotes which object to save.Available options are:\n",
    "#' \\itemize{\n",
    "#'   \\item \"seurat\": converted the \"seurat object\" to the h5 file.\n",
    "#'   \\item \"singlecellexperiment\": converted the \"singlecellexperiment object\" to the h5 file.\n",
    "#'   \\item \"monocle\": converted the \"monocle3 object\" to the h5 file.\n",
    "#' }\n",
    "#' @param assay.name 'assay.name' is used to flag the data type. Defualt is \"RNA\". Available options are:\n",
    "#' \\itemize{\n",
    "#'   \\item \"RNA\": this is the scRNA-seq data.\n",
    "#'   \\item \"spatial\": this is the spatial data.}\n",
    "#' @param save.graphs Default is TRUE , determing whether to save the graph(cell-cell similarity network).\n",
    "#'   seurat graph is different from scanpy graph. Their relationship are set {\"distances\": \"knn\", \"connectivities\": \"snn\"} roughly.\n",
    "#' @param save.scale Default is FALSE, determint whether to save the scale.data(dense matrix)\n",
    "#' @export\n",
    "#'\n",
    "write_h5 <- function(data, file, object.type = 'seurat', assay.name = 'RNA', save.graphs = TRUE, save.scale=FALSE){\n",
    "  if(object.type == 'seurat'){\n",
    "    seurat_write_h5(seurat = data, file = file, assay.name = assay.name, save.graphs = save.graphs, save.scale = save.scale)\n",
    "  }\n",
    "  if(object.type == 'singlecellexperiment'){\n",
    "    sce_write_h5(sce = data, file = file, assay.name = assay.name)\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "#--- read_h5 part\n",
    "\n",
    "#' The read the some data in the h5 file\n",
    "#'\n",
    "#' @param file the h5 file\n",
    "#' @param groups the h5 groups\n",
    "#' @importFrom hdf5r H5File h5attr\n",
    "#' @return The list contains the core data for single-cell information\n",
    "#'\n",
    "#' @export\n",
    "read_h5part <- function(file, groups){\n",
    "  h5 <- H5File$new(filename = file, mode = 'r')\n",
    "  h5_list <- list()\n",
    "  for(h in groups){\n",
    "    h5_list[[h]] <- switch(h, data = to_data_(h5),\n",
    "                               obs = to_obs_(h5),\n",
    "                               var = to_var_(h5),\n",
    "                               dimR = seurat.to_dimr_(h5),\n",
    "                               layers = to_layers_(h5),\n",
    "                               graphs = to_graphs_(h5),\n",
    "                               spatial = seurat.to_spatial_(h5),\n",
    "                               uns = to_uns_(h5))\n",
    "  }\n",
    "  return(h5_list)\n",
    "}\n",
    "\n",
    "#' reading the h5ad file\n",
    "#'\n",
    "#' @param file the file name\n",
    "#' @param assay_name the assay_name\n",
    "#' @param target.object the target.object name\n",
    "#'\n",
    "#' @export\n",
    "read_h5ad <-function(file,\n",
    "                     assay_name,\n",
    "                     target.object){\n",
    "\n",
    "  rfile = paste0(system.file(package = 'dior'), '/python/diorPy.py')\n",
    "  if(file.exists(rfile)){\n",
    "    rfile = rfile\n",
    "  }else{\n",
    "    rfile = paste0(system.file(package = 'dior'), '/inst/python/diorPy.py')\n",
    "  }\n",
    "  py_script = paste0('python ', rfile, ' -f ', file, ' -a ', assay_name)\n",
    "  system(py_script)\n",
    "  data = read_h5(file = gsub('.h5ad', '_tmp.h5', file), target.object = target.object)\n",
    "  return(data)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "31ea0e5c-4e63-4eda-a641-21be151d77fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (file, assay_name, target.object) \n",
       "{\n",
       "<span style=white-space:pre-wrap>    rfile = paste0(system.file(package = \"dior\"), \"/python/diorPy.py\")</span>\n",
       "<span style=white-space:pre-wrap>    if (file.exists(rfile)) {</span>\n",
       "<span style=white-space:pre-wrap>        rfile = rfile</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        rfile = paste0(system.file(package = \"dior\"), \"/inst/python/diorPy.py\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    py_script = paste0(\"python \", rfile, \" -f \", file, \" -a \", </span>\n",
       "<span style=white-space:pre-wrap>        assay_name)</span>\n",
       "<span style=white-space:pre-wrap>    system(py_script)</span>\n",
       "<span style=white-space:pre-wrap>    data = read_h5(file = gsub(\".h5ad\", \"_tmp.h5\", file), target.object = target.object)</span>\n",
       "<span style=white-space:pre-wrap>    return(data)</span>\n",
       "}</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (file, assay\\_name, target.object) \n",
       "\\{\n",
       "    rfile = paste0(system.file(package = \"dior\"), \"/python/diorPy.py\")\n",
       "    if (file.exists(rfile)) \\{\n",
       "        rfile = rfile\n",
       "    \\}\n",
       "    else \\{\n",
       "        rfile = paste0(system.file(package = \"dior\"), \"/inst/python/diorPy.py\")\n",
       "    \\}\n",
       "    py\\_script = paste0(\"python \", rfile, \" -f \", file, \" -a \", \n",
       "        assay\\_name)\n",
       "    system(py\\_script)\n",
       "    data = read\\_h5(file = gsub(\".h5ad\", \"\\_tmp.h5\", file), target.object = target.object)\n",
       "    return(data)\n",
       "\\}\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (file, assay_name, target.object) \n",
       "{\n",
       "    rfile = paste0(system.file(package = \"dior\"), \"/python/diorPy.py\")\n",
       "    if (file.exists(rfile)) {\n",
       "        rfile = rfile\n",
       "    }\n",
       "    else {\n",
       "        rfile = paste0(system.file(package = \"dior\"), \"/inst/python/diorPy.py\")\n",
       "    }\n",
       "    py_script = paste0(\"python \", rfile, \" -f \", file, \" -a \", \n",
       "        assay_name)\n",
       "    system(py_script)\n",
       "    data = read_h5(file = gsub(\".h5ad\", \"_tmp.h5\", file), target.object = target.object)\n",
       "    return(data)\n",
       "}\n",
       "```"
      ],
      "text/plain": [
       "function(file,\n",
       "                     assay_name,\n",
       "                     target.object){\n",
       "\n",
       "  rfile = paste0(system.file(package = 'dior'), '/python/diorPy.py')\n",
       "  if(file.exists(rfile)){\n",
       "    rfile = rfile\n",
       "  }else{\n",
       "    rfile = paste0(system.file(package = 'dior'), '/inst/python/diorPy.py')\n",
       "  }\n",
       "  py_script = paste0('python ', rfile, ' -f ', file, ' -a ', assay_name)\n",
       "  system(py_script)\n",
       "  data = read_h5(file = gsub('.h5ad', '_tmp.h5', file), target.object = target.object)\n",
       "  return(data)\n",
       "}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (data, file, object.type = \"seurat\", assay.name = \"RNA\", \n",
       "<span style=white-space:pre-wrap>    save.graphs = TRUE, save.scale = FALSE) </span>\n",
       "{\n",
       "<span style=white-space:pre-wrap>    if (object.type == \"seurat\") {</span>\n",
       "<span style=white-space:pre-wrap>        seurat_write_h5(seurat = data, file = file, assay.name = assay.name, </span>\n",
       "<span style=white-space:pre-wrap>            save.graphs = save.graphs, save.scale = save.scale)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    if (object.type == \"singlecellexperiment\") {</span>\n",
       "<span style=white-space:pre-wrap>        sce_write_h5(sce = data, file = file, assay.name = assay.name)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "}</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (data, file, object.type = \"seurat\", assay.name = \"RNA\", \n",
       "    save.graphs = TRUE, save.scale = FALSE) \n",
       "\\{\n",
       "    if (object.type == \"seurat\") \\{\n",
       "        seurat\\_write\\_h5(seurat = data, file = file, assay.name = assay.name, \n",
       "            save.graphs = save.graphs, save.scale = save.scale)\n",
       "    \\}\n",
       "    if (object.type == \"singlecellexperiment\") \\{\n",
       "        sce\\_write\\_h5(sce = data, file = file, assay.name = assay.name)\n",
       "    \\}\n",
       "\\}\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (data, file, object.type = \"seurat\", assay.name = \"RNA\", \n",
       "    save.graphs = TRUE, save.scale = FALSE) \n",
       "{\n",
       "    if (object.type == \"seurat\") {\n",
       "        seurat_write_h5(seurat = data, file = file, assay.name = assay.name, \n",
       "            save.graphs = save.graphs, save.scale = save.scale)\n",
       "    }\n",
       "    if (object.type == \"singlecellexperiment\") {\n",
       "        sce_write_h5(sce = data, file = file, assay.name = assay.name)\n",
       "    }\n",
       "}\n",
       "```"
      ],
      "text/plain": [
       "function(data, file, object.type = 'seurat', assay.name = 'RNA', save.graphs = TRUE, save.scale=FALSE){\n",
       "  if(object.type == 'seurat'){\n",
       "    seurat_write_h5(seurat = data, file = file, assay.name = assay.name, save.graphs = save.graphs, save.scale = save.scale)\n",
       "  }\n",
       "  if(object.type == 'singlecellexperiment'){\n",
       "    sce_write_h5(sce = data, file = file, assay.name = assay.name)\n",
       "  }\n",
       "}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>function (file, target.object = \"seurat\", assay.name = \"RNA\") \n",
       "{\n",
       "<span style=white-space:pre-wrap>    if (target.object == \"seurat\") {</span>\n",
       "<span style=white-space:pre-wrap>        data &lt;- seurat_read_h5(file = file, assay.name = assay.name)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else if (target.object == \"singlecellexperiment\") {</span>\n",
       "<span style=white-space:pre-wrap>        data &lt;- sce_read_h5(file = file, assay.name = assay.name)</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    else {</span>\n",
       "<span style=white-space:pre-wrap>        stop(\"The output object must be specified\")</span>\n",
       "<span style=white-space:pre-wrap>    }</span>\n",
       "<span style=white-space:pre-wrap>    return(data)</span>\n",
       "}</code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "function (file, target.object = \"seurat\", assay.name = \"RNA\") \n",
       "\\{\n",
       "    if (target.object == \"seurat\") \\{\n",
       "        data <- seurat\\_read\\_h5(file = file, assay.name = assay.name)\n",
       "    \\}\n",
       "    else if (target.object == \"singlecellexperiment\") \\{\n",
       "        data <- sce\\_read\\_h5(file = file, assay.name = assay.name)\n",
       "    \\}\n",
       "    else \\{\n",
       "        stop(\"The output object must be specified\")\n",
       "    \\}\n",
       "    return(data)\n",
       "\\}\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "function (file, target.object = \"seurat\", assay.name = \"RNA\") \n",
       "{\n",
       "    if (target.object == \"seurat\") {\n",
       "        data <- seurat_read_h5(file = file, assay.name = assay.name)\n",
       "    }\n",
       "    else if (target.object == \"singlecellexperiment\") {\n",
       "        data <- sce_read_h5(file = file, assay.name = assay.name)\n",
       "    }\n",
       "    else {\n",
       "        stop(\"The output object must be specified\")\n",
       "    }\n",
       "    return(data)\n",
       "}\n",
       "```"
      ],
      "text/plain": [
       "function(file, target.object = 'seurat', assay.name = 'RNA'){\n",
       "  if(target.object == 'seurat'){\n",
       "    data <- seurat_read_h5(file = file, assay.name = assay.name)\n",
       "  }\n",
       "  else if(target.object == 'singlecellexperiment'){\n",
       "    data <- sce_read_h5(file = file, assay.name = assay.name)\n",
       "  }\n",
       "  else{\n",
       "    stop('The output object must be specified')\n",
       "  }\n",
       "  return(data)\n",
       "}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "read_h5ad ##调用pyhoon paste0(system.file(package = 'dior'), '/inst/python/diorPy.py') 的 wite_h 生成 tmp.h5  然受读入tmp.h5 s生成 rds的seurat 对象\n",
    "write_h5 ##自身生成rds 生成 h5\n",
    "read_h5  ##tmp.h5 生成 rds 的seurat对象"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a782f961-262d-40a8-927a-258e850f3d67",
   "metadata": {},
   "outputs": [],
   "source": [
    "####调用 python 让h5ad 生成tmp.h5\n",
    "  rfile = paste0(system.file(package = 'dior'), '/python/diorPy.py')\n",
    "  if(file.exists(rfile)){\n",
    "    rfile = rfile\n",
    "  }else{\n",
    "    rfile = paste0(system.file(package = 'dior'), '/inst/python/diorPy.py')\n",
    "  }\n",
    "\n",
    "  file=\"a_test.h5ad\"\n",
    "  assay_name=\"RNA\"\n",
    "  target.object=\"seurat\"\n",
    "  \n",
    "  py_script = paste0('python ', rfile, ' -f ', file, ' -a ', assay_name)\n",
    "  system(py_script)\n",
    "\n",
    "# ## /home/zhulab/miniconda3/envs/R4/lib/R/library/dior/python/diorPy.py\n",
    "# # python diopy \n",
    "\n",
    "# import argparse\n",
    "# import scanpy as sc\n",
    "# from diopy import output\n",
    "# import re\n",
    "\n",
    "# # set the parser\n",
    "# parser = argparse.ArgumentParser(description='R read the h5ad ', epilog=\"diorpy -f file.h5ad -t seurat\")\n",
    "# parser.add_argument('-f', '--file', dest = 'file', help = 'The h5ad file', type = str, required = True)\n",
    "# parser.add_argument('-a', '--assay_name', dest = 'assay_name', help = 'The data type assay', type = str, required = True)\n",
    "# args = parser.parse_args()\n",
    "\n",
    "# # save the h5 file\n",
    "# adata = sc.read(args.file)\n",
    "# output.write_h5(adata = adata,file=re.sub('.h5ad', '_tmp.h5', args.file), assay_name=args.assay_name)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c40b16f4-fc66-45ea-b4eb-9a0a16d5ee34",
   "metadata": {},
   "outputs": [],
   "source": [
    "####调用 python 让 tmp.h5 写入 h5ad\n",
    "  rfile = paste0(system.file(package = 'dior'), '/python/diorPyOh5ad.py')\n",
    "  if(file.exists(rfile)){\n",
    "    rfile = rfile\n",
    "  }else{\n",
    "    rfile = paste0(system.file(package = 'dior'), '/inst/python/diorPyOh5ad.py')\n",
    "  }\n",
    "\n",
    "  file=\"a_test_tmp.h5\"\n",
    "  assay_name=\"RNA\"\n",
    "  target.object=\"seurat\"\n",
    "  outfile=\"a_test_tmp.h5ad\"\n",
    "  \n",
    "  py_script = paste0('python ', rfile, ' -f ', file, ' -o ', outfile, ' -a ', assay_name)\n",
    "  system(py_script)\n",
    "\n",
    "#!python /home/zhulab/miniconda3/envs/R4/lib/R/library/dior/python/diorPyOh5ad.py -f a_test_tmp.h5 -o a_test_tmp.h5ad -a RNA\n",
    "## 自定义文件/home/zhulab/miniconda3/envs/R4/lib/R/library/dior/python/diorPyOh5ad.py\n",
    "# import argparse\n",
    "# import scanpy as sc\n",
    "# from diopy import input\n",
    "# import re\n",
    "\n",
    "# # set the parser\n",
    "# parser = argparse.ArgumentParser(description='R write the h5ad ', epilog=\"diorpy -f file.h5ad -t seurat\")\n",
    "# parser.add_argument('-f', '--file', dest = 'file', help = 'The h5 file', type = str, required = True)\n",
    "# parser.add_argument('-o', '--outfile', dest = 'outfile', help = 'The h5ad file name', type = str, required = True)\n",
    "# parser.add_argument('-a', '--assay_name', dest = 'assay_name', help = 'The data type assay', type = str, required = True)\n",
    "# args = parser.parse_args()\n",
    "\n",
    "# # save the h5 file\n",
    "# #re.sub('.h5ad', '_tmp.h5', args.file)\n",
    "# h5a=input.read_h5(file=args.file, assay_name=args.assay_name)\n",
    "# h5a.write(args.outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c18ea4bf-ba8b-47b6-9c49-64085ae1d244",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'python /home/zhulab/miniconda3/envs/R4/lib/R/library/dior/python/diorPyOh5ad.py -f a_test_tmp.h5 -o a_test_tmp.h5ad -a RNA'"
      ],
      "text/latex": [
       "'python /home/zhulab/miniconda3/envs/R4/lib/R/library/dior/python/diorPyOh5ad.py -f a\\_test\\_tmp.h5 -o a\\_test\\_tmp.h5ad -a RNA'"
      ],
      "text/markdown": [
       "'python /home/zhulab/miniconda3/envs/R4/lib/R/library/dior/python/diorPyOh5ad.py -f a_test_tmp.h5 -o a_test_tmp.h5ad -a RNA'"
      ],
      "text/plain": [
       "[1] \"python /home/zhulab/miniconda3/envs/R4/lib/R/library/dior/python/diorPyOh5ad.py -f a_test_tmp.h5 -o a_test_tmp.h5ad -a RNA\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "py_script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0421e718-64fa-4abc-8a97-99b90eaba400",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "13364 features across 3669 samples within 1 assay \n",
       "Active assay: RNA (13364 features, 0 variable features)\n",
       " 2 dimensional reductions calculated: pca, umap"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "assay_name=\"RNA\"\n",
    "target.object=\"seurat\"\n",
    "adata=read_h5ad(\"a_test.h5ad\",assay_name,\n",
    "                     target.object) ##调用pyhoon paste0(system.file(package = 'dior'), '/inst/python/diorPy.py') 的 wite_h 生成 tmp.h5  然受读入tmp.h5 s生成 rds的seurat 对象\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1db16c14-7959-41c6-bd09-ba5633a395b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data=adata\n",
    "file=\"atesth5_tmp.h5\"\n",
    "write_h5(data, file, object.type = 'seurat', assay.name = 'RNA', save.graphs = TRUE, save.scale=FALSE) ##自身生成rds 生成 h5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3d4b0d5c-60f1-45a9-b3fd-a1ee5cd60385",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "13364 features across 3669 samples within 1 assay \n",
       "Active assay: RNA (13364 features, 0 variable features)\n",
       " 2 dimensional reductions calculated: pca, umap"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seuratobj=read_h5(file, target.object = 'seurat', assay.name = 'RNA')  ##tmp.h5 生成 rds 的seurat对象\n",
    "seuratobj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56893f4f-8e4e-4813-ab01-5bf3fa5c6d62",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f4e06d3-7562-4bbb-a7cb-b2b9b1057fa4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
